# -*- coding: utf-8 -*-
from setuptools import setup

package_dir = \
{'': 'src'}

packages = \
['himut']

package_data = \
{'': ['*']}

install_requires = \
['argparse>=1.4.0,<2.0.0',
 'biopython>=1.78,<2.0',
 'cyvcf2>=0.30.18,<0.31.0',
 'matplotlib>=3.5.3,<4.0.0',
 'mizani>=0.8.1,<0.9.0',
 'natsort>=8.0.0,<9.0.0',
 'numpy>=1.20.2,<2.0.0',
 'plotnine>=0.9.0,<0.10.0',
 'psutil>=5.8.0,<6.0.0',
 'pyfastx>=0.8.4,<0.9.0',
 'pyproject-toml>=0.0.10,<0.0.11',
 'pysam>=0.18.0,<0.19.0',
 'pytabix>=0.1,<0.2']

entry_points = \
{'console_scripts': ['himut = himut.__main__:main']}

setup_kwargs = {
    'name': 'himut',
    'version': '1.0.4',
    'description': 'himut: single molecule somatic single-base substitution detection using PacBio CCS reads',
    'long_description': '## <a name="started"></a>Getting started\n\n```sh\n## pip installation\npip install himut\n\n## download and install the latest release\nwget https://github.com/sjin09/himut/archive/refs/tags/v1.0.2.tar.gz\ntar -zxvf v1.0.3.tar.gz\ncd himut-1.0.3\nbash install.sh\n\n## use miniamp2 and samtools to align, sort (and merge) PacBio CCS read alignments\nminimap2 "@RG\\tSM:sample" -ax map-hifi --cs ref.fa pacbio.ccs.fastq.gz | samtools sort -o aln.sorted.bam # SM tag must be provided to retrieve sample ID\nsamtools merge *.sorted.bam | samtools sort -o aln.mergeSorted.bam - ## if there are multiple BAM files, merge and sort the BAM files \nsamtools view -bh -F 0x900 aln.sorted.bam > aln.primary_alignments.sorted.bam # select primary alignments\nsamtools index aln.sorted.bam\nsamtools index aln.primary_alignments.sorted.bam \n\n## germline mutation detection and haplotype phasing\ndeepvariant.simg /opt/deepvariant/bin/run_deepvariant --model_type=PACBIO --ref ref.fa --reads=aln.primary_alignments.sorted.bam --output_vcf=germline.vcf ## use deepvariant to call germline mutations \nbgzip -c germline.vcf > germline.vcf.bgz ## bgzip compress\ntabix -p vcf germline.vcf.bgz ## tabix index\nhimut phase --bam aln.primary_alignments.sorted.bam --vcf germline.vcf.bgz -o germline.phased.vcf ## phsae\nbgzip -c germline.phased.vcf  > germline.phased.vcf.bgz ## bgzip compress\ntabix -p vcf germline.phased.vcf.bgz ## tabix index\n\n## somatic mutation detection in human samples\n\n### use himut to call somatic mutations in human samples\nhimut call -i aln.primary_alignments.sorted.bam -o somatic.vcf  # somatic mutation detection\nhimut call -i aln.primary_alignments.sorted.bam --phased_vcf germline.phased.vcf.bgz -o somatic.vcf --phase # haplotype phased somatic mutation detection\nhimut call -i aln.primary_alignments.sorted.bam --common_snps ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.decomposed.normalised.common_snps.vcf.bgz --panel_of_normals b37.himut.pon.sbs.dbs.vcf.bgz -o somatic.vcf # provide a VCF file with common SNPs or a panel of normal VCF file\nhimut call -i aln.primary_alignments.sorted.bam --phased_vcf germline.phased.vcf.bgz --common_snps ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.decomposed.normalised.common_snps.vcf.bgz --panel_of_normals b37.himut.pon.sbs.dbs.vcf.bgz -o somatic.vcf --phase # somatic mutation detection with the highest sensitivity \nbgzip -c somatic.vcf > somatic.vcf.bgz\ntabix -p vcf somatic.vcf.bgz\n\n### normalise SBS96 counts ## use the same parameters as that used for somatic mutation detection \nhimut normcounts --bam aln.primary_alignments.sorted.bam --ref ref.fa --sbs somatic.vcf.bgz -o somatic.normcounts.tsv\nhimut normcounts --bam aln.primary_alignments.sorted.bam --ref ref.fa --sbs somatic.vcf.bgz --phased_vcf germline.phased.vcf.bgz -o somatic.normcounts.tsv --phase\nhimut normcounts --bam aln.primary_alignments.sorted.bam --ref ref.fa --sbs somatic.vcf.bgz --common_snps ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.decomposed.normalised.common_snps.vcf.bgz --panel_of_normals b37.himut.pon.sbs.dbs.vcf.bgz -o somatic.normcounts.tsv\nhimut normcounts --bam aln.primary_alignments.sorted.bam --ref ref.fa --sbs somatic.vcf.bgz --common_snps ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.decomposed.normalised.common_snps.vcf.bgz --panel_of_normals b37.himut.pon.sbs.dbs.vcf.bgz --phased_vcf germline.phased.vcf.bgz -o somatic.normcounts.tsv --phase\n\n## somatic mutation detection in non-human samples\n\n### use himut to call somatic mutations in non-human samples\nhimut call -i aln.primary_alignments.sorted.bam --vcf germline.vcf.bgz -o somatic.vcf --non_human_sample \nhimut call -i aln.primary_alignments.sorted.bam --vcf germline.vcf.bgz --phased_vcf germline.phased.vcf.bgz -o somatic.vcf --phase --non_human_sample \nbgzip -c somatic.vcf > somatic.vcf.bgz\ntabix -p vcf somatic.vcf.bgz\n\n## normalise SBS96 counts\nhimut normcounts --bam aln.primary_alignments.sorted.bam --ref ref.fa --vcf germline.vcf.bgz --sbs somatic.vcf.bgz -o somatic.normcounts.tsv\nhimut normcounts --bam aln.primary_alignments.sorted.bam --ref ref.fa --vcf germline.vcf.bgz --phased_vcf germline.phased.vcf.bgz --sbs somatic.vcf.bgz -o somatic.normcounts.tsv\n```\n\n## Table of Contents\n\n- [Getting Started](#started)\n- [Users\' Guide](#uguide)\n  - [Installation](#install)\n  - [minimap2 CCS read alignment](#alignment)\n  - [General usage](#general)\n    - [Somatic mutation detection](#som)\n    - [Haplotype phased somatic mutation detection](#phasedsom)\n    - [SBS96 count calculation](#SBS96) \n    - [SBS96 counts normalisation](#normcounts)\n    - [Somatic mutation detection across the Tree of Life](#tol)\n  - [Panel of Normal VCF file](#pon)\n  - [VCF file with common SNPs](#pop)\n  - [Citing himut](#cite)\n  - [Limitations](#limits)\n\n## <a name="uguide"></a>Users\' Guide\n\nhimut (high-fidelity mutation) is a somatic mutation caller that leverages the base accuracy and read length of Pacific Biosciences (PacBio) CCS reads to call single molecule somatic single-base substitutions (SBS). \n\nA single-read mismatch bewteen a sample and the reference genome is derived from library errors, sequencing errors or somatic mutations. If sequencing error rate is lower than the somatic mutation rate, single-molecule somatic mutation detection is possible. On the other hand, if sequencing error rate is higher than the somatic mutation rate, but if the number of accumulated somatic mutations is higher than the number of sequencing errors, which is a constant, somatic mutational process associated with the sample can be directly observed from the mutational spectrum.\n\n### <a name="install"></a>Installation\n\nUse pip to install himut:\n\n```sh\npip install himut\n```\n\nDownload and install the latest release:\n\n```sh\nwget https://github.com/sjin09/himut/archive/refs/tags/v1.0.3.tar.gz\ntar -zxvf v1.0.2.tar.gz\ncd himut-1.0.2\nbash install.sh\n```\n\nYou can download and install the source code with the latest developments that might contain code still under development.\n\n```sh\ngit clone https://github.com/sjin09/himut \ncd himut\n/bin/bash install.sh \n````\n\nInstallation through Bioconda will be supported in the future.\n\n#### <a name="alignment"></a>minimap2 CCS read alignment\n\nCurrently, himut only accepts minimap2 CCS read alignment SAM/BAM file with a `--cs` tag, like the `cigar` string provides information about the matches and mismatches between the read and the reference genome. pbmm2 and ngmlr does not provide the `--cs` tag and as a result, pbmm2 and ngmlr SAM/BAM files are incompatible with himut. The `--cs` tag is favored over the `cigar` string because it offers a more elegant representation of the differences between the read and the reference genome (source: https://lh3.github.io/2018/03/27/the-history-the-cigar-x-operator-and-the-md-tag).\n\n```sh\nminimap2 "@RG\\tSM:sample" -ax map-hifi --cs ref.fa pacbio.ccs.fastq.gz | samtools sort -o aln.sorted.bam # SM tag must be provided to retrieve sample ID\nsamtools merge *.sorted.bam | samtools sort -o aln.mergeSorted.bam - # merge read alignments\nsamtools view -bh -F 0x900 aln.mergeSorted.bam > aln.primary_alignments.mergeSorted.bam # select primary alignments\nsamtools index aln.primary_alignments.sorted.bam\nsamtools index aln.primary_alignments.mergeSorted.bam \n```\n\n### <a name="general"></a>General Usage\n\nCurrently, himut only accepts minimap2 generated BAM files, as `--cs` tag is required for somatic mutation detection. The CCS read alignments must be sorted, \\(merged\\), compressed and indexed using SAMtools.\n\n#### <a name="som"></a>Somatic mutation detection\n\nHimut accepts as input a BAM file with CCS read alignments and returns a VCF file with somatic mutations. \n\n```sh\nhimut call -i sample.primary_alignments.sorted.bam -o sample.somatic.vcf  \n```\n\nhimut can optionally accept a VCF file with common SNPs and/or a panel of normal VCF file to filter substitutions derived from genomic DNA contamination and/or from non-canonical error modes, respectively.\n\n```sh\nhimut call -i aln.primary_alignments.sorted.bam --common_snps ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.decomposed.normalised.common_snps.vcf.bgz --panel_of_normals b37.himut.pon.sbs.dbs.vcf.bgz -o somatic.vcf # provide a VCF file with common SNPs or a panel of normal VCF file\n```\n\n#### <a name="phasedsom"></a>Haplotype phased somatic mutation detection \n\nHimut leverages CCS read length and base accuracy to haplotype phase germline mutations. CCS reads have an average read length between 10kb and 20kb and consequently, a CCS read can span mulitple heterozaygous single nucleotide polymorphisms (hetSNPs). Himut identifies haplotype consistent hetSNPs that are connected to construct haplotype blocks. In the VCF file, haplotype phased hetSNPs in the same haplotype block have the same phase set (PS), which is the position of the first hetSNP in the haplotype block. \n\n```sh\nhimut phase --bam aln.primary_alignments.sorted.bam --vcf germline.vcf.bgz -o germline.phased.vcf ## phase\nbgzip -c germline.phased.vcf  > germline.phased.vcf.bgz ## bgzip compress\ntabix -p vcf germline.phased.vcf.bgz ## tabix index\n```\n\nHimut can take the haplotype phased VCF file as an additional input to haplotype phase CCS reads and limit somatic mutation detection to haplotype phased CCS reads. \n\n```sh\nhimut call -i aln.primary_alignments.sorted.bam --phased_vcf germline.phased.vcf.bgz -o somatic.vcf --phase # haplotype phased somatic mutation detection\nhimut call -i aln.primary_alignments.sorted.bam --phased_vcf germline.phased.vcf.bgz --common_snps ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.decomposed.normalised.common_snps.vcf.bgz --panel_of_normals b37.himut.pon.sbs.dbs.vcf.bgz -o somatic.vcf --phase # somatic mutation detection with the highest sensitivity \nbgzip -c somatic.vcf > somatic.vcf.bgz\ntabix -p vcf somatic.vcf.bgz\n```\n\n#### <a name="SBS96"></a>SBS96 count calculation\n\nThere are 6 possible substitutions in the pyrimidine sequence context (C>A, C>G, C>T, T>A, T> and T>G) and 16 possible trinucleotide sequence contexts for each substitution, creating the SBS96 classification system (6 * 16 = 96). \n\nHimut accepts as input a VCF file with somatic mutations and a reference genome FASTA file, classifies each substitution and returns a file with SBS96 counts.\n\n```sh\nhimut sbs96 -i somatic.vcf --ref ref.fa -o somatic.sbs96.tsv \n```\n\n#### <a name="normcounts"></a>SBS96 count normalisation \n\nBefore mutational signature extraction and analysis, SBS96 count needs to be normalised based on the trinucleotide frequencines in the reference genome FASTA file and the callable trinucleotide frequencies in CCS reads. This is because number of callable bases, the CCS bases from which somatic mutation detection could have been performed, is unique to each sample and each sequencing run. To calculate the number of callable CCS bases, SBS96 count normalisation parameters needs to be identical to the somatic mutation detection parameters, as the parameters changes the number of callable CCS bases. \n\n```sh\nhimut normcounts --bam aln.primary_alignments.sorted.bam --ref ref.fa --common_snps ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.decomposed.normalised.common_snps.vcf.bgz --panel_of_normals b37.himut.pon.sbs.dbs.vcf.bgz --sbs somatic.vcf.bgz -o somatic.normcounts.tsv ## unphased normalisation\nhimut normcounts --bam aln.primary_alignments.sorted.bam --ref ref.fa --phased_vcf germline.phased.vcf.bgz --common_snps ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.decomposed.normalised.common_snps.vcf.bgz --panel_of_normals b37.himut.pon.sbs.dbs.vcf.bgz --sbs somatic.vcf.bgz -o somatic.normcounts.tsv --phase ## phased normalisation\n```\n\n#### <a name="tol"></a>Somatic mutation detection across the Tree of Life\n\nHimut enable somatic mutation detection from bulk normal tissue, agnostic of species and clonality, and can be used to detect somatic mutations across the Tree of Life. We used himut to call somatic mutations from Darwin Tree of Life (DToL) project samples. Himut requires germline SNP prior to calculate the genotype quality score for each germline muation and because the germline SNP prior is unknown, a deepvariant VCF file needs to be provided to calculate the germline SNP prior. How the germline SNP prior is calculated depends on whether CCS reads and reference genome is derived from the same sample or from a different sample. \n\nTo indicate to himut that germline SNP prior needs to be calculated, provide `--non_human_sample` as a parameter. In addition, if the CCS reads and reference genome are derived from the same sample, provide `--reference_sample` as a parameter. In addition, because the DToL project sequences and assembles one sample per species, a VCF file with common SNPs and a Panel of Normal VCF file is not available. \n\n```sh\n## unphased somatic mutation detection and SBS96 count normalisation where CCS reads and reference genome is derived from the same sample\nhimut call -i aln.primary_alignments.sorted.bam --vcf germline.vcf.bgz -o somatic.vcf --non_human_sample --reference_sample \nhimut normcounts --bam aln.primary_alignments.sorted.bam --ref ref.fa --vcf germline.vcf --phased_vcf germline.phased.vcf.bgz --non_human_sample --reference_sample\n\n## phased somatic mutation detection and SBS96 count normalisation where CCS reads and reference genome is derived from the same sample\nhimut call -i aln.primary_alignments.sorted.bam --vcf germline.vcf.bgz --phased_vcf germline.phased.vcf.bgz -o somatic.vcf --phase --non_human_sample --reference_sample\nhimut normcounts --bam aln.primary_alignments.sorted.bam --ref ref.fa --vcf germline.vcf --phased_vcf germline.phased.vcf.bgz --non_human_sample --reference_sample\n```\n\nIf the CCS reads and reference genome are derived from a different sample, use the following commands.\n\n```sh\n## unphased somatic mutation detection and SBS96 count normalisation where CCS reads and reference genome is derived from a different sample\nhimut call -i aln.primary_alignments.sorted.bam --vcf germline.vcf.bgz -o somatic.vcf --non_human_sample\nhimut normcounts --bam aln.primary_alignments.sorted.bam --ref ref.fa --vcf germline.vcf --phased_vcf germline.phased.vcf.bgz --non_human_sample\n\n## phased somatic mutation detection and SBS96 count normalisation where CCS reads and reference genome is derived from a different sample\nhimut call -i aln.primary_alignments.sorted.bam --vcf germline.vcf.bgz --phased_vcf germline.phased.vcf.bgz -o somatic.vcf --phase --non_human_sample\nhimut normcounts --bam aln.primary_alignments.sorted.bam --ref ref.fa --vcf germline.vcf --phased_vcf germline.phased.vcf.bgz --non_human_sample\n```\n\n#### <a name="pon"></a>Panel of Normal VCF file\n\nA Panel of Normal VCF file is used to remove false positive substitutions resulting from non-canonical errors. To create a Panel of Normal VCF file, please provide himut with a minimap2 BAM file from an unrelated individual from the same species and the `--create_panel_of_normal` parameter. Himut will use less stringent thresholds to maximise the number of substitutions called from the sample and if the same found in a sample of interest, the substitution is not considered as a somatic mutation. \n\n```sh\nhimut call -i sample.aln.primary_alignments.sorted.bam -o sample.pon.vcf --create_panel_of_normal\n```\n\nWe prepared a Panel of Normal VCF file for public use Google Drive ((https://drive.google.com/drive/folders/14-cSTBocwdyIemXTExuhW-mnTJb87SLT?usp=sharing).\n\n#### <a name="pop"></a>VCF file with common SNPs\n\nWe selected common SNPs (>1% minor allele frequency) from the 1000 Genomes Project Phase 3 VCF file (https://drive.google.com/drive/folders/14-cSTBocwdyIemXTExuhW-mnTJb87SLT?usp=sharing). \n\n\n### <a name="limits"></a>Limitations\n\n- Himut does not support Revio CCS reads, which has BQ score that ranges from Q1 to Q40. \n- PacBio CCS base accuracy is the limiting factor for somatic mutation calling using himut. PacBio circular consensus sequence (pbccs) algorithm calculates the base quality score depending on the dinucleotide sequence context and the number of supporting subread bases. Our research suggests that the base quality score does not correctly reflect the true base accuracy and that the consensus sequence caller does not account for the different substitution error rates in different trinucleotide sequence contexts. We discuss how this issue can be mitigated in our manuscript.\n',
    'author': 'Sangjin Lee',
    'author_email': 'sl17@sanger.ac.uk',
    'maintainer': 'None',
    'maintainer_email': 'None',
    'url': 'https://github.com/sjin09/himut',
    'package_dir': package_dir,
    'packages': packages,
    'package_data': package_data,
    'install_requires': install_requires,
    'entry_points': entry_points,
    'python_requires': '>=3.8,<3.10',
}


setup(**setup_kwargs)

